File: android/src/main/kotlin/com/nocorps/in_app_browser/InAppBrowserPlugin.kt
package com.nocorps.in_app_browser

import android.content.Context
import android.view.View
import android.webkit.WebView
import android.webkit.WebViewClient
import android.webkit.WebChromeClient
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class InAppBrowserPlugin: FlutterPlugin, MethodChannel.MethodCallHandler {
  private lateinit var channel: MethodChannel

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(binding.binaryMessenger, "com.nocorps.in_app_browser")
    channel.setMethodCallHandler(this)
    binding.platformViewRegistry.registerViewFactory(
      "com.nocorps.in_app_browser/webview", WebViewFactory(binding.binaryMessenger))
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

class WebViewFactory(private val messenger: BinaryMessenger) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
  override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
    val creationParams = args as Map<String?, Any?>?
    return FlutterWebView(context, messenger, viewId, creationParams)
  }
}

class FlutterWebView(
  context: Context,
  messenger: BinaryMessenger,
  id: Int,
  creationParams: Map<String?, Any?>?
) : PlatformView, MethodChannel.MethodCallHandler {
  private val webView: WebView = WebView(context)
  private val methodChannel: MethodChannel = MethodChannel(messenger, "com.nocorps.in_app_browser/webview_$id")

  init {
    methodChannel.setMethodCallHandler(this)
    
    // Configure WebView
    webView.settings.apply {
      javaScriptEnabled = true
      domStorageEnabled = true
      databaseEnabled = true
      useWideViewPort = true
      loadWithOverviewMode = true
      allowFileAccess = true
      allowContentAccess = true
    }
    
    webView.webViewClient = object : WebViewClient() {
      override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
        methodChannel.invokeMethod("onLoadingStateChanged", mapOf("isLoading" to true))
        methodChannel.invokeMethod("onUrlChanged", mapOf("url" to url))
      }

      override fun onPageFinished(view: WebView, url: String) {
        methodChannel.invokeMethod("onLoadingStateChanged", mapOf("isLoading" to false))
        methodChannel.invokeMethod("onUrlChanged", mapOf("url" to url))
        webView.evaluateJavascript("document.title") { title ->
          val processedTitle = title?.replace("\"", "") ?: ""
          methodChannel.invokeMethod("onTitleChanged", mapOf("title" to processedTitle))
        }
      }
    }
    
    webView.webChromeClient = object : WebChromeClient() {
      override fun onProgressChanged(view: WebView, progress: Int) {
        methodChannel.invokeMethod("onProgressChanged", mapOf("progress" to progress / 100.0))
      }
    }
    
    // Load initial URL if provided
    val initialUrl = creationParams?.get("initialUrl") as? String
    if (initialUrl != null) {
      webView.loadUrl(initialUrl)
    }
    
    // Apply settings
    val settings = creationParams?.get("settings") as? Map<String, Any>
    if (settings != null) {
      applySettings(settings)
    }
  }

  private fun applySettings(settings: Map<String, Any>) {
    webView.settings.apply {
      settings["javaScriptEnabled"]?.let { javaScriptEnabled = it as Boolean }
      settings["domStorageEnabled"]?.let { domStorageEnabled = it as Boolean }
      settings["databaseEnabled"]?.let { databaseEnabled = it as Boolean }
      settings["useWideViewPort"]?.let { useWideViewPort = it as Boolean }
      settings["allowFileAccess"]?.let { allowFileAccess = it as Boolean }
      settings["allowContentAccess"]?.let { allowContentAccess = it as Boolean }
      settings["loadWithOverviewMode"]?.let { loadWithOverviewMode = it as Boolean }
    }
  }

  override fun getView(): View = webView

  override fun dispose() {
    methodChannel.setMethodCallHandler(null)
    webView.destroy()
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "loadUrl" -> {
        val url = call.argument<String>("url")
        if (url != null) {
          webView.loadUrl(url)
          result.success(null)
        } else {
          result.error("MISSING_PARAMS", "URL is required", null)
        }
      }
      "goBack" -> {
        if (webView.canGoBack()) {
          webView.goBack()
        }
        result.success(null)
      }
      "goForward" -> {
        if (webView.canGoForward()) {
          webView.goForward()
        }
        result.success(null)
      }
      "reload" -> {
        webView.reload()
        result.success(null)
      }
      "canGoBack" -> {
        result.success(webView.canGoBack())
      }
      "canGoForward" -> {
        result.success(webView.canGoForward())
      }
      "getCurrentUrl" -> {
        result.success(webView.url)
      }
      "getTitle" -> {
        result.success(webView.title)
      }
      "evaluateJavascript" -> {
        val javascript = call.argument<String>("javascript")
        if (javascript != null) {
          webView.evaluateJavascript(javascript) { value ->
            result.success(value)
          }
        } else {
          result.error("MISSING_PARAMS", "JavaScript code is required", null)
        }
      }
      else -> result.notImplemented()
    }
  }
}

File: ios/Classes/InAppBrowserPlugin.swift
import Flutter
import UIKit
import WebKit

public class InAppBrowserPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let factory = InAppBrowserViewFactory(messenger: registrar.messenger())
    registrar.register(factory, withId: "com.nocorps.in_app_browser/webview")
  }
}

class InAppBrowserViewFactory: NSObject, FlutterPlatformViewFactory {
  private let messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
    return InAppBrowserView(
      frame: frame,
      viewIdentifier: viewId,
      arguments: args as? [String: Any],
      messenger: messenger
    )
  }

  public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }
}

class InAppBrowserView: NSObject, FlutterPlatformView, WKNavigationDelegate, WKUIDelegate {
  private let webView: WKWebView
  private let methodChannel: FlutterMethodChannel
  private var progressObservation: NSKeyValueObservation?

  init(frame: CGRect, viewIdentifier viewId: Int64, arguments args: [String: Any]?, messenger: FlutterBinaryMessenger) {
    let configuration = WKWebViewConfiguration()
    webView = WKWebView(frame: frame, configuration: configuration)
    methodChannel = FlutterMethodChannel(name: "com.nocorps.in_app_browser/webview_\(viewId)", binaryMessenger: messenger)
    
    super.init()
    
    // Set delegates
    webView.navigationDelegate = self
    webView.uiDelegate = self
    
    // Apply settings if provided
    if let settings = args?["settings"] as? [String: Any] {
      applySettings(settings)
    }
    
    // Load initial URL if provided
    if let initialUrl = args?["initialUrl"] as? String {
      if let url = URL(string: initialUrl) {
        webView.load(URLRequest(url: url))
      }
    }
    
    // Set up progress observer
    progressObservation = webView.observe(\.estimatedProgress, options: [.new]) { [weak self] webView, _ in
      guard let self = self else { return }
      self.methodChannel.invokeMethod("onProgressChanged", arguments: ["progress": webView.estimatedProgress])
    }
    
    // Set up method channel handler
    methodChannel.setMethodCallHandler { [weak self] (call, result) in
      guard let self = self else { return }
      self.handleMethodCall(call, result: result)
    }
  }

  func view() -> UIView {
    return webView
  }
  
  private func applySettings(_ settings: [String: Any]) {
    if let javaScriptEnabled = settings["javascriptEnabled"] as? Bool {
      webView.configuration.preferences.javaScriptEnabled = javaScriptEnabled
    }
    
    // For other WKWebView settings
    if let contentMode = settings["useWideViewPort"] as? Bool, contentMode {
      webView.configuration.preferences.setValue(true, forKey: "allowsContentJavaScript")
    }
  }
  
  private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "loadUrl":
      if let args = call.arguments as? [String: Any], let urlString = args["url"] as? String {
        if let url = URL(string: urlString) {
          webView.load(URLRequest(url: url))
          result(nil)
        } else {
          result(FlutterError(code: "INVALID_URL", message: "The URL is invalid", details: nil))
        }
      } else {
        result(FlutterError(code: "MISSING_PARAMS", message: "URL is required", details: nil))
      }
      
    case "goBack":
      if webView.canGoBack {
        webView.goBack()
      }
      result(nil)
      
    case "goForward":
      if webView.canGoForward {
        webView.goForward()
      }
      result(nil)
      
    case "reload":
      webView.reload()
      result(nil)
      
    case "canGoBack":
      result(webView.canGoBack)
      
    case "canGoForward":
      result(webView.canGoForward)
      
    case "getCurrentUrl":
      result(webView.url?.absoluteString ?? "")
      
    case "getTitle":
      result(webView.title ?? "")
      
    case "evaluateJavascript":
      if let args = call.arguments as? [String: Any], let javaScript = args["javascript"] as? String {
        webView.evaluateJavaScript(javaScript) { (value, error) in
          if let error = error {
            result(FlutterError(code: "JS_ERROR", message: error.localizedDescription, details: nil))
          } else {
            if let stringValue = value as? String {
              result(stringValue)
            } else if let value = value {
              result("\(value)")
            } else {
              result("")
            }
          }
        }
      } else {
        result(FlutterError(code: "MISSING_PARAMS", message: "JavaScript code is required", details: nil))
      }
      
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  // MARK: - WKNavigationDelegate
  
  func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
    methodChannel.invokeMethod("onPageStarted", arguments: ["url": webView.url?.absoluteString ?? ""])
  }
  
  func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
    methodChannel.invokeMethod("onPageFinished", arguments: ["url": webView.url?.absoluteString ?? ""])
    methodChannel.invokeMethod("onTitleChanged", arguments: ["title": webView.title ?? ""])
  }
  
  func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
    methodChannel.invokeMethod("onPageFinished", arguments: ["url": webView.url?.absoluteString ?? ""])
  }
  
  func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
    methodChannel.invokeMethod("onPageFinished", arguments: ["url": webView.url?.absoluteString ?? ""])
  }
  
  // MARK: - Cleanup
  
  deinit {
    progressObservation?.invalidate()